prboom2/src/dsda/zipfile.c

```diff
@@ -0,0 +1,134 @@
+//
+// Copyright(C) 2022 by Pierre Wendling
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// DESCRIPTION:
+//	DSDA zipfile support using libzip
+//
+
+#include <stdio.h>
+#include <zip.h>
+
+#include "lprintf.h"
+#include "m_file.h"
+#include "z_zone.h"
+
+#include "dsda/utility.h"
+
+static char *dsda_JoinPath(const char *prefix, const char *suffix) {
+  int joined_path_length;
+  char *joined_path;
+
+  joined_path_length = snprintf(NULL, 0, "%s%s", prefix, suffix) + 1;
+  joined_path = (char *)Z_Malloc(joined_path_length);
+  if (joined_path == NULL) {
+    lprintf(LO_ERROR, "dsda_JoinPath: Failed to allocate joined path.\n");
+    return NULL;
+  }
+  snprintf(joined_path, joined_path_length, "%s%s", prefix, suffix);
+  return joined_path;
+}
+
+static void dsda_CreateSubDir(const char *directory, const char *subdirectory) {
+  char *full_path;
+  int result;
+
+  full_path = dsda_JoinPath(directory, subdirectory);
+  if (full_path == NULL) {
+    return;
+  }
+  result = M_MakeDir(full_path, false);
+  Z_Free(full_path);
+  if (result != 0) {
+    lprintf(LO_ERROR, "dsda_CreateSubDir: Failed to create subdirectory %s in %s.\n", subdirectory, directory);
+  }
+}
+
+static FILE *dsda_CreateFile(const char *directory, const char *filename) {
+  FILE *file_handle;
+  char *full_path;
+
+  full_path = dsda_JoinPath(directory, filename);
+  if (full_path == NULL) {
+    return NULL;
+  }
+  file_handle = M_OpenFile(full_path, "wb");
+  Z_Free(full_path);
+  return file_handle;
+}
+
+static void dsda_WriteContentToFile(zip_file_t *input_file, FILE *dest_file, zip_uint64_t data_size) {
```

Comment:
Take a look at `dboolean M_WriteFile(char const *name, const void *source, size_t length)` in `m_file`. You can allocate the buffer, call `zip_fread`, and then call `M_WriteFile` using the path, rather than creating a file separately. `M_WriteFile` also manages removing the file if there's an error writing it. You should probably also check if the return value matches the expected length and raise an error if not.

Reply:
I went ahead and used a fixed-sized buffer instead as suggested by @bkoropoff.